File: 3568-minimum-moves-to-clean-the-classroom/3568-minimum-moves-to-clean-the-classroom.py
from collections import deque

class Solution:
    def minMoves(self, classroom: list[str], energy: int) -> int:
        m, n = len(classroom), len(classroom[0])
        start_r = start_c = -1
        litters = []
        
        # Locate starting position 'S' and store coordinates of all litter cells 'L'
        for r in range(m):
            for c in range(n):
                cell = classroom[r][c]
                if cell == 'S':
                    start_r, start_c = r, c
                elif cell == 'L':
                    litters.append((r, c))
                    
        total_litter = len(litters)
        target_mask = (1 << total_litter) - 1 # Bitmask representing all litters collected
        
        # Map grid coordinates to unique bit index (0 to total_litter - 1)
        litter_idx = {(r, c): i for i, (r, c) in enumerate(litters)}
        
        # max_energy[r][c][mask] tracks max remaining energy for a (position, mask) state
        max_energy = [[[-1] * (1 << total_litter) for _ in range(n)] for _ in range(m)]
        
        # Queue state: (row, col, current_energy, collected_mask)
        queue = deque([(start_r, start_c, energy, 0)])
        max_energy[start_r][start_c][0] = energy
        
        moves = 0
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        
        # Level-order BFS traversal guarantees finding the minimum number of moves
        while queue:
            for _ in range(len(queue)):
                r, c, e, mask = queue.popleft()
                
                if mask == target_mask: 
                    return moves # Goal reached: all litter collected
                    
                if e == 0: 
                    continue # Cannot move without remaining energy
                    
                for dr, dc in dirs:
                    nr, nc = r + dr, c + dc
                    
                    # Boundary check and obstacle check ('X')
                    if 0 <= nr < m and 0 <= nc < n and classroom[nr][nc] != 'X':
                        next_e = e - 1
                        next_mask = mask
                        cell = classroom[nr][nc]
                        
                        if cell == 'L':
                            next_mask |= (1 << litter_idx[(nr, nc)]) # Set litter bit as collected
                        elif cell == 'R':
                            next_e = energy                         # Reset energy to max capacity
                            
                        # Prune state unless it yields strictly higher remaining energy
                        if next_e > max_energy[nr][nc][next_mask]:
                            max_energy[nr][nc][next_mask] = next_e
                            queue.append((nr, nc, next_e, next_mask))
                            
            moves += 1 # Increment step count per BFS level
            
        return -1 # Return -1 if unreachable

File: 3568-minimum-moves-to-clean-the-classroom/3568-minimum-moves-to-clean-the-classroom.java
import java.util.*;

class Solution {
    public int minMoves(String[] classroom, int energy) {
        int m = classroom.length, n = classroom[0].length();
        int startR = -1, startC = -1;
        List<int[]> litters = new ArrayList<>();
        // Locate starting position 'S' and store coordinates of all litter cells 'L'
        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                char ch = classroom[r].charAt(c);
                if (ch == 'S') { startR = r; startC = c; }
                else if (ch == 'L') litters.add(new int[]{r, c});
            }
        }
        int totalLitter = litters.size();
        int targetMask = (1 << totalLitter) - 1; // Bitmask representing all litters collected (all 1s)
        // Map grid coordinates to unique bit index (0 to totalLitter - 1)
        int[][] litterIdx = new int[m][n];
        for (int i = 0; i < totalLitter; i++) {
            int[] pos = litters.get(i);
            litterIdx[pos[0]][pos[1]] = i;
        }
        // maxEnergy[r][c][mask] tracks max remaining energy for a (position, mask) state
        int[][][] maxEnergy = new int[m][n][1 << totalLitter];
        for (int[][] layer : maxEnergy) {
            for (int[] row : layer) Arrays.fill(row, -1);
        }
        // Queue stores primitive bit-packed ints: [mask (10b) | energy (8b) | col (5b) | row (5b)]
        Deque<Integer> queue = new ArrayDeque<>();
        queue.offer(pack(startR, startC, energy, 0));
        maxEnergy[startR][startC][0] = energy;
        int moves = 0;
        int[][] dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        // Level-order BFS traversal guarantees finding the minimum number of moves
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                int curr = queue.poll();
                int r = curr & 31;            // Unpack row (bits 0-4)
                int c = (curr >> 5) & 31;      // Unpack col (bits 5-9)
                int e = (curr >> 10) & 255;    // Unpack energy (bits 10-17)
                int mask = curr >> 18;         // Unpack collected mask (bits 18+)
                if (mask == targetMask) return moves; // Goal reached: all litter collected
                if (e == 0) continue;          // Cannot move without remaining energy
                for (int[] d : dirs) {
                    int nr = r + d[0], nc = c + d[1];
                    // Boundary check and obstacle check ('X')
                    if (nr < 0 || nr >= m || nc < 0 || nc >= n || classroom[nr].charAt(nc) == 'X') continue;
                    int nextEnergy = e - 1, nextMask = mask;
                    char cell = classroom[nr].charAt(nc);
                    if (cell == 'L') nextMask |= (1 << litterIdx[nr][nc]); // Set litter bit as collected
                    else if (cell == 'R') nextEnergy = energy;           // Reset energy to max capacity
                    // Prune state unless it yields strictly higher remaining energy
                    if (nextEnergy > maxEnergy[nr][nc][nextMask]) {
                        maxEnergy[nr][nc][nextMask] = nextEnergy;
                        queue.offer(pack(nr, nc, nextEnergy, nextMask));
                    }
                }
            }
            moves++; // Increment step count per BFS level
        }
        return -1; // Return -1 if unreachable
    }
    // Packs (row, col, energy, mask) into a single 32-bit int to eliminate garbage collection
    private int pack(int r, int c, int e, int mask) {
        return r | (c << 5) | (e << 10) | (mask << 18);
    }
}